### ISSUE TITLE
Disable tests by default

### ISSUE THREAD

**Issue by tigertv** [tigertv/big-integer#4]
> It would be better if there was an option to disable tests while importing the library, like that - [mekhovyyekulaki@4dcf417](https://github.com/mekhovyyekulaki/BeeNum/commit/4dcf4178d792e41389fc3cd4f05b99e24488fa97)
> They get interferred with the project's tests which make the library less comfortable to use. 

(from [PR#3](https://github.com/tigertv/BeeNum/pull/3))

**Comment by saxbophone** [comment]
For CMake-based projects, I usually check if my lib is being used as a sub-project with this hack:

```cmake
if(NOT DEFINED PROJECT_NAME)
    set(LIB_IS_SUBPROJECT OFF)
else()
    set(LIB_IS_SUBPROJECT ON)
endif()

# call to project() MUST come after!
project(...)
```

and just use an implicit option `ENABLE_TESTING` to switch on or off the tests, in combination with making sure we're not a sub-project:

```cmake
if(ENABLE_TESTS AND NOT LIB_IS_SUBPROJECT)
    message(STATUS "Unit Tests Enabled")
    add_subdirectory(tests)
    enable_testing()
endif()
```

One can then enable the tests explicitly by passing `-DENABLE_TESTS=ON` on the cmake command-line at generate time.

Personally, I go a step further and detect when in debug mode and enable tests automatically in that case, disable them automatically in release mode (but they can be manually reënabled by passing the flag as mentioned above).

**Comment by tigertv** [comment]
@saxbophone,  OK, Great!
Could you add a pull request?

**Comment by saxbophone** [comment]
> @saxbophone, OK, Great! Could you add a pull request?

I could. What behaviour do you want?
-  [ ] Do you want the tests to not run by default **unless** explicitly enabled by the user?
- [x] Or do you want the tests to run by default in Debug mode but not Release mode?

**Comment by tigertv** [comment]
The case is the following:
Tests are disabled in Release mode, and they are enabled in Debug mode.
But user can change the behavior by passing the option ENABLE_TESTING.

What do you think?

**Comment by saxbophone** [comment]
> The case is the following: Tests are disabled in Release mode, and they are enabled in Debug mode. But user can change the behavior by passing the option ENABLE_TESTING.
> 
> What do you think?

This is what I use for my own projects and it works well for me, is very convenient for CI and for manually building the tests in release mode if one's ever worried about compiler optimisations breaking them.

I can do this no problem!

**Comment by tigertv** [comment]
@saxbophone , 
Is it hard?

**Comment by saxbophone** [comment]
> @saxbophone , Is it hard?

No, I've just been busy doing other stuff and haven't got round to it yet. You will be able to see when I start working on it as I will fork your project.

**Comment by tigertv** [comment]
OK.

**Comment by tigertv** [comment]
Somebody lies on the Internet again...

**Comment by saxbophone** [comment]
> Somebody lies on the Internet again...

You know, I really was going to do this for you when I found the time for it, but if that's your attitude then you know what, never mind. You can do it yourself.

The code samples I've already given in comments here should be a good enough starting point for you.

**Comment by tigertv** [comment]
[Paroles, paroles](https://www.youtube.com/watch?v=bDqHEArTo-Q) :)
